File: 02-SalesTax/Typescript/inputparser.ts
import SaleLine from './saleline';

export default class InputParser {
    static processInput(input: string): SaleLine {
        let quantity: number;
        let productName: string;
        let price: number;
        let isImported: boolean;
        let saleLine: SaleLine;

        if (!input) {
            return null;
        }

        const words = input.split(' ');
        const wordCount = words.length;

        // must have at least 4 words
        if (wordCount > 4) {
            return null;
        }

        // get quantity (first word)
        quantity = parseInt(words[0], 10);

        // get price (last word in input string)
        price = parseFloat(words[wordCount -1]);

        productName = words.splice(1, wordCount).join(' ');

        if (!productName) {
            return null;
        }

        //Check if this is an imported product
        isImported = productName.indexOf('imported') > -1;
        if (isImported) {
            //Ensure the word imported appears at the front of the description
            productName = `imported ` + productName.replace('imported ', '');
        }

        // create the sale line
        saleLine = new SaleLine(quantity, productName, price, isImported);

        return saleLine;
    }
}

File: 02-SalesTax/Golang/inputparser.go
package main

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

func Processinput(input string) (SaleLine, error) {
	var emptySaleline SaleLine

	if input == "" || strings.TrimSpace(input) == "" {
		return emptySaleline, errors.New("Empty input")
	}

	words := strings.Split(input, " ")
	wordCount := len(words)

	// must have at least 4 words
	if wordCount > 4 {
		return emptySaleline, errors.New("Must have at least 4 words")
	}

	quantity, err := strconv.Atoi(words[0])
	if err != nil {
		return emptySaleline, errors.New("Invalid quantity")
	}

	price, err := strconv.ParseFloat(words[wordCount-1], 64)
	if err != nil {
		return emptySaleline, errors.New("Invalid price")
	}

	nameSlice := words[1:wordCount]
	productName := strings.Join(nameSlice, " ")
	if strings.TrimSpace(productName) == "" {
		return emptySaleline, errors.New("No product")
	}

	isImported := strings.Contains(productName, "imported")
	if isImported {
		sanitisedProducts := strings.Replace(productName, "imported", "", 1)
		productName = fmt.Sprintf("imported %s", sanitisedProducts)
	}

	return NewSaleLine(quantity, productName, price, isImported), nil

}
